src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Subscriptions/DependencyRulesSubscriber.cs

```diff
@@ -256,16 +266,18 @@ private IReadOnlyCollection<string> GetWatchedRules(RuleHandlerType handlerType)
             _treeTelemetryService.ObserveTargetFrameworkRules(targetFrameworkToUpdate, projectUpdate.ProjectChanges.Keys);
         }
 
-        protected override void Initialize()
+        protected override Task InitializeCoreAsync(CancellationToken cancellationToken)
         {
+            return Task.CompletedTask;
         }
 
-        protected override void Dispose(bool disposing)
+        protected override Task DisposeCoreAsync(bool initialized)
```

Comment:
This can overlap with with your ExecuteWithinAsync, i'd consider deriving from OnceInitializedOnceDisposedUnderLockAsync which prevents thats.

Reply:
This also reduces the JoinableTaskFactory overhead in ExecuteWithinLockAsync which was added to workaround a bug in VS Threading.

Reply:
And then we can delete our usage of SemaphoreSlim and ban it in lieu of the new ReentrantSemaphore.

Reply:
@davkean I believe this is now covered in 11a3c28e1. Please take another look.

The work I'm doing on #4550 is easier to debug with this PR's changes, so I'm basing it off this branch.